DataFormats/PatCandidates/src/PackedCandidate.cc

```diff
@@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
 }
 
 void pat::PackedCandidate::unpack() const {
-    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
+    float pt = MiniFloatConverter::float16to32(packedPt_);
+    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
+    double phi = int16_t(packedPhi_)*3.2f/std::numeric_limits<int16_t>::max() + shift*3.2/std::numeric_limits<int16_t>::max();
```

Comment:
On 6/11/15 12:05 PM, Dinko Ferencek wrote:
> In DataFormats/PatCandidates/src/PackedCandidate.cc
> <https://github.com/cms-sw/cmssw/pull/9554#discussion_r32243557>:
> 
>> @@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
>>  }
>>  
>>  void pat::PackedCandidate::unpack() const {
>> -    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
>> +    float pt = MiniFloatConverter::float16to32(packedPt_);
>> +    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
>> +    double phi = int16_t(packedPhi_)*3.2f/std::numeric_limits<int16_t>::max() + shift*3.2/std::numeric_limits<int16_t>::max();
> 
> By construction, the degridification cannot be done at the packing
> stage, since it is the packing that introduces the grid (the packing
> process involves mapping to integers so there are no fractional integers
> available). The degridification can only be introduced at the unpacking
> stage.


As I suggested earlier on the HN, packing of a non-linear f(x) to
integers f_i
and then unpacking to x = fInverse (f_i) will give x_i = fInverse(f_i)
values not on the regular grid.

This approach is stable under pack->unpack cycle for all the x_i values.

For eta, it could be packing of theta (in email I mentioned eta**1.00001)

For phi, it could be packing of cos(phi) [+ 1 bit for pi offset]




> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9554/files#r32243557>.
>